haystack/components/samplers/top_p.py

```diff
@@ -99,44 +106,80 @@ def run(self, documents: List[Document], top_p: Optional[float] = None):
         top_p_indices = torch.where(torch.BoolTensor(condition))[0]
 
         # Map the selected indices back to their original indices
-        original_indices = sorted_indices[top_p_indices]
-        selected_docs = [documents[i.item()] for i in original_indices]
+        selected_docs = [sorted_documents[i.item()] for i in top_p_indices]
 
-        # If low p resulted in no documents being selected, then
-        # return at least one document
-        if not selected_docs:
+        if self.min_top_k and len(selected_docs) < self.min_top_k:
+            selected_docs = sorted_documents[: self.min_top_k]
+
+        # If low p resulted in no documents being selected, then return at least one document
+        if len(selected_docs) == 0:
             logger.warning(
                 "Top-p sampling with p={top_p} resulted in no documents being selected. "
-                "Returning the document with the highest similarity score.",
+                "Returning the document with the highest score.",
                 top_p=top_p,
             )
-            highest_prob_indices = torch.argsort(probs, descending=True)
-            selected_docs = [documents[int(highest_prob_indices[0].item())]]
+            selected_docs = [sorted_documents[0]]
 
         return {"documents": selected_docs}
 
-    def _collect_scores(self, documents: List[Document]) -> List[float]:
+    def _get_doc_score_from_meta(self, doc: Document) -> Optional[float]:
+        """
+        Get the score of a document from its metadata.
+
+        :param doc: Document object.
+        :return: Score of the document.
         """
-        Collect the scores from the documents' metadata.
+        score: Optional[float] = doc.meta.get(self.score_field)  # type: ignore
```

Comment:
I kept this one since it complains that self.score_field could be `None` even though it's not possible since this function is only called when self.score_field is not None.